Add redhat_manifest module

sean797 opened the pull request:
This uses the RHSM API to modify, create or download a Satellite manifest from the Red Hat portal.

akofink commented:
I appreciate the initiative, but I'm not sure this module belongs in the foreman + plugins set of modules. Perhaps it would be better to add this subscription management application logic to core Ansible? Perhaps the functionality could be included in the [redhat_subscription module](http://docs.ansible.com/ansible/latest/redhat_subscription_module.html).

sean797 commented:
@akofink I think it does belong here as AFAIK the only reason someone has to generate a manifest is Katello (or Satellite) ?

ekohl commented on modules/rhsm_manifest.py:
```diff
@@ -0,0 +1,264 @@
+    headers = {'accept': accept_header,
+               'content-type': 'application/json'}
+    resp, info = fetch_url(module, url, json.dumps(data), headers, method)
+    return resp, info
```
This makes the code below unreachable. Is that intentional?

ekohl commented on modules/rhsm_manifest.py:
```diff
@@ -0,0 +1,264 @@
+    subs = get_subs(module, manifest)
+    if subs:
+        if module.params['state'] == 'present':
+            sub_quantity = sum(map(lambda s: s['quantity'], subs))
```
Is this easier to read? ```sum(s['quantity'] for s in subs)```

ekohl commented on modules/rhsm_manifest.py:
```diff
@@ -0,0 +1,264 @@
+    path = "/subscription/consumers/%s/entitlements" % (manifest['uuid'])
+    resp, info = fetch_portal(module, path, 'GET')
+    all_subs = json.loads(resp.read())
+    subs = filter(lambda s: s['pool']['id'] == module.params['pool_id'], all_subs)
```
Note that in Python 3 filter has become a generator which means you can consume it only once. That may lead to very subtle breakage. I'd advise a list comprehension: ```[s for s in all_subs if s['pool']['id'] == module.params['pool_id']]``` or wrapping the above in ```list()```

ekohl commented on modules/rhsm_manifest.py:
```diff
@@ -0,0 +1,264 @@
+
+def remove_sub(module, manifest, sub):
+    path = "/subscription/consumers/%s/entitlements/%s" % (manifest['uuid'], sub['id'])
+    resp, info = fetch_portal(module, path, 'DELETE')
```
Unused variables. I'd either remove them or return something.

ekohl commented on modules/rhsm_manifest.py:
```diff
@@ -0,0 +1,264 @@
+    path = "/subscription/consumers/%s/export" % (manifest['uuid'])
+    try:
+        resp, info = fetch_portal(module, path, 'GET', accept_header='application/zip')
+        f = open(module.params['path'], 'w')
```
Using the ```with``` syntax will autoclose it even if an exception is raised:
```python
with open(module.params['path'], 'w') as f:
    while True:
        data = data.read(65536)
        if not data:
            break
        f.write(data)
```

ekohl commented on modules/rhsm_manifest.py:
```diff
@@ -0,0 +1,264 @@
+            f.write(data)
+        f.close()
+    except Exception:
+        e = get_exception()
```
Any reason you're not using ```except Exception as e```? Is this an ansible thing?

sean797 replied:
👍 `get_exception()` is an ansible thing, https://github.com/theforeman/foreman-ansible-modules/pull/71/files#diff-88ef7a0ec70e64151c3a125a04cb3b36R114

ekohl commented on modules/rhsm_manifest.py:
```diff
@@ -0,0 +1,264 @@
+    if module.params['path']:
+        export_manifest(module, manifest)
+
+    changed = man_changed if man_changed else sub_changed
```
I think ```change = man_changed or sub_changed``` should work just as well but you should be careful that if the module has no ```pool_id``` that ```sub_changed``` is not defined.

ekohl commented on modules/rhsm_manifest.py:
```diff
@@ -0,0 +1,264 @@
+                remove_sub(module, manifest, subs[0])
+                changed = True
+                subs = get_subs(module, manifest)
+                sub_quantity = sum(map(lambda s: s['quantity'], subs))
```
This can be rewritten like above.

ekohl commented on modules/rhsm_manifest.py:
```diff
@@ -0,0 +1,264 @@
+        attr = 'name'
+    if module.params['uuid']:
+        attr = 'uuid'
+    manifest = filter(lambda m: m[attr] == module.params[attr], manifests)
```
Same as ```get_subs```

sean797 commented:
Thanks @ekohl 👍 updated

ekohl commented on modules/rhsm_manifest.py:
```diff
@@ -0,0 +1,272 @@
+        export_manifest(module, manifest)
+
+    changed = man_changed
+    if module.params['pool_id'] and not changed:
```
I'd simplify this and add ```else: sub_changed = False``` to other if block. Since you added the ```and manifest, you need to duplicate the logic here to but being able to count on it being present makes it much easier. Then you can go back to ```changed = man_changed or sub_changed```

sean797 commented:
Yea that's quite a bit simpler! Thanks @ekohl

akofink commented:
@sean797 All other modules in this repo use the Foreman/Katello API, which is why I think it makes sense to have `foreman_*`/`katello_*` modules. This one would hit the RHSM API, which makes me think it should be in a `rhsm_*` module. I'd at least like to get some other opinions on it, maybe from @ehelms or @ekohl?

ekohl commented:
I have no opinion on it. I see your point but the module is named ```rhsm_*```. If there is no other rhsm module, I'd be fine with including it here rather than have the overhead of a separate repo.

akofink commented:
@ekohl Also remember that Ansible modules are usually added directly to the ansible/ansible repo, and that's the ultimate goal of this repo, that way users can install Ansible and be able to use these modules out of the box.

ekohl commented:
That's fair, but if including it here for a bit exposes it to more testing before it's included in ansible then fine by me.

akofink commented:
Yeah, I'd ultimately be fine with it staying in this repo; just wanted to see what others thought. Thanks for the quick response!

ehelms commented:
With the US holidays I have not gotten to this yet but I'll weigh in
beginning of next week.

On Nov 22, 2017 7:27 PM, "sean797" <notifications@github.com> wrote:

> This uses the RHSM API to modify, create or download a Satellite manifest
> from the Red Hat portal.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/theforeman/foreman-ansible-modules/pull/71
> Commit Summary
>
>    - Add rhsm_manifest module
>
> File Changes
>
>    - *M* README.md
>    <https://github.com/theforeman/foreman-ansible-modules/pull/71/files#diff-0>
>    (1)
>    - *A* modules/rhsm_manifest.py
>    <https://github.com/theforeman/foreman-ansible-modules/pull/71/files#diff-1>
>    (264)
>    - *R* test/data/katello_manifest.json
>    <https://github.com/theforeman/foreman-ansible-modules/pull/71/files#diff-2>
>    (0)
>    - *R* test/demo/katello_manifest.yml
>    <https://github.com/theforeman/foreman-ansible-modules/pull/71/files#diff-3>
>    (0)
>    - *A* test/demo/rhsm_manifest.yml
>    <https://github.com/theforeman/foreman-ansible-modules/pull/71/files#diff-4>
>    (42)
>    - *M* test/demo/server_vars.yml.example
>    <https://github.com/theforeman/foreman-ansible-modules/pull/71/files#diff-5>
>    (6)
>
> Patch Links:
>
>    - https://github.com/theforeman/foreman-ansible-modules/pull/71.patch
>    - https://github.com/theforeman/foreman-ansible-modules/pull/71.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/theforeman/foreman-ansible-modules/pull/71>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAR58wwJbZzcxHR0esvX_NfBELSrLf-kks5s5LwJgaJpZM4QoJ1a>
> .
>

mdellweg commented:
While i am quite indifferent about the inclusion-or-not question, i think that renaming the test data files breaks expectations of the (not yet so) automatic tests we have in the pipeline.

akofink commented:
@mdellweg Those tests aren't yet in the repo, so it's mostly a matter of which PR gets merged first (probably this one). When I rebase the testing PR I'll move the test files over.

ehelms commented on modules/rhsm_manifest.py:
```diff
@@ -0,0 +1,271 @@
+            'type': "satellite",
+            'owner': module.params['rhsm_owner'],
+            'facts': {'distributor_version': 'sat-6.0',
+                      'system.certificate_version': '3.2'}}
```
Should either of these be configurable to protect against future changes in the portal?

sean797 replied:
Maybe..  

I have no idea what `system.certificate_version` actually does?
`distributor_version` should probably be, though I'm not sure what API we need to call to modify it on an already existing manifest, added a TODO comment for now.

Most of the APIs I got form https://github.com/Katello/hammer-cli-csv/commit/648163df7fe0dc0ead3ea0186eaba6d10d7f7b96. Maybe @thomasmckay has some insights?

ehelms commented on modules/rhsm_manifest.py:
```diff
@@ -0,0 +1,271 @@
+    pool_id: XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+    quantity: 7
+    state: present
+    validate_certs: false
```
If I understand correctly, this example ensures that the manifest would have 7 of pool XXXXX?

sean797 replied:
Correct 👍

ehelms commented on modules/rhsm_manifest.py:
```diff
@@ -0,0 +1,271 @@
+    quantity: 10
+    state: present
+    validate_certs: false
+    path: /root/manifest.zip
```
If I understand correctly, this example ensures that the manifest would have 10 of pool XXXXXXXX AND would then download that manifest and store it?

sean797 replied:
Correct 👍

sean797 commented:
If you have got other suggestions on the name, I'm open to ideas.. I wouldn't focus too much on the reason I am doing it, others might have different reasons to do? Though I do not think there is currently any other reason to get a manifest.

Yes, the normal workflow would be attach subs, probably using `with_items`, then download it using another task. e.g I am currently using:

```yaml
- name: Attach subs
  rhsm_manifest:
    name: "{{ configure_katello_manifest_name }}"
    username: "{{ configure_rhsm_username }}"
    password: "{{ configure_rhsm_password }}"
    pool_id: "{{ item.pool_id }}"
    quantity: "{{ item.quantity }}"
    state: "{{ item.state }}"
    validate_certs: "{{ configure_rhsm_verify_ssl }}"
  with_items: "{{ configure_katello_subscriptions }}"
  when: configure_katello_subscriptions is defined
  register: subs

- name: Download manifest
  rhsm_manifest:
    name: "{{ configure_katello_manifest_name }}"
    username: "{{ configure_rhsm_username }}"
    password: "{{ configure_rhsm_password }}"
    validate_certs: "{{ configure_rhsm_verify_ssl }}"
    path: "{{ ansible_env['HOME'] }}/{{ configure_katello_manifest_name }}.zip"
  when:
    - configure_katello_manifest_name is defined
    - subs.changed
```


`UUID` is completely optional, I added it because name is not unique, it is possible to have 2 manifests in the portal with the same name (if name is supplied we just assume the first one)

ehelms commented:
I suppose you could then also chain this with (https://github.com/theforeman/foreman-ansible-modules/blob/master/modules/katello_manifest.py) to upload the exported manifest or initiate a refresh? 

I think this all sounds great and gives users a nice way until the Katello UI provides support to manage a manifest as code. Given there is a module named `redhat_subscription` in Ansible core, would `redhat_manifest` be too over the top to stay in line with that naming?

sean797 commented:
@ehelms Yes! That's exactly what I do, take a look at https://github.com/sean797/ansible-role-configure-katello/blob/master/tasks/manifest.yml if you are interested. 

I agree with the name, updated!

ehelms commented:
Given others reviewed, I'd like to get an additional functionality/code review ACK from either @akofink or @ekohl

ekohl commented on test/test_playbooks/rhsm_manifest.yml:
```diff
@@ -0,0 +1,42 @@
+---
```
Should this also be named ```redhat_manifest.yml``` to match the module name?

sean797 commented:
Thanks again @ekohl. Updated.

akofink commented:
When adding a redhat manifest with 1 of POOL_ID, then updating the quantity attached to 3, I get this:

![image](https://user-images.githubusercontent.com/761923/33607487-4aeed2b0-d98f-11e7-929b-eaa38eb08b05.png)

Is this expected/desired? Why not a single listing of 3?

akofink commented on modules/redhat_manifest.py:
```diff
@@ -0,0 +1,274 @@
+        default: present
+        choices:
+            - present
+            - absent
```
`state` should be the redhat_manifest state, don't you think? I expect that if I use `state: absent`, the manifest will be removed from access.redhat.com.

sean797 replied:
If it was the manifest state, then you wouldnt be able to remove a sub from a manifest.

Do you think the name should be `redhat_manifest_sub` or something else? Or maybe if state == absent and there are no other subs in it then we delete the manifest?

akofink replied:
Yeah, you're not going to like this response, but I think the task should represent the state of the manifest. Perhaps it'd be better to add pools/quantities in a dict like so (state in this example is the manifest state, not subscriptions):

```yaml
  - name: "Create {{ manifest_name }} manifest and add 5 rhel subs and ensure 0 rhev subs"
    redhat_manifest:
      name: "{{ manifest_name }}"
      username: "{{ rhsm_username }}"
      password: "{{ rhsm_password }}"
      subscriptions:
        - pool_id: "{{ rhsm_rhel_pool_id }}"
          quantity: 5
        - pool_id: "{{ rhsm_rhev_pool_id }}"
          quantity: 0
      state: present
      validate_certs: "{{ rhsm_validate_certs }}"
```

Sorry I didn't catch this earlier. We can discuss it more and let others give their input, but afaik every `state` param in the modules in the repo refer to the module's object itself (i.e. organization, product, etc.)

akofink replied:
@mdellweg @ekohl  I'd like to get your opinions here.

ekohl replied:
Without knowing too much details I think I agree with @akofink. state is at the manifest level so it should ensure the state of a manifest.

sean797 replied:
@akofink I did originally think about doing it like that, but I think its better to use `with_items` if you need to add multiple subs, I think having complex dict type options takes away form the simplicity of Ansible.

akofink replied:
@sean797 You can still implement it in such a way that you can add each subscription one at a time if you prefer; however, users wishing to give their APIs a break will probably want to make a single API call for all their subscriptions.

> complex dict type options

I think Ansible is simple because it uses yaml instead of a DSL. I think users are quite comfortable with more complex yaml (and this is just a list of dicts), as it's intention is to be human-readable. The complexity more comes in the module implementation itself rather than what users will see, which I don't think is an issue.

mdellweg replied:
I am really no expert for redhat subscriptions. But maybe that gives me the naïve viewpoint you are aiming for.
I agree, that the `state` parameter should really be used for the entity described in the module name.
However, if i understand the discussion right, and `quantity: 0` is not enough to delete a subscription, another `state: present|absent` on the subscription level could make the addition/deletion of subscriptions from the manifest more explicit.

akofink replied:
> another state: present|absent on the subscription level could make the addition/deletion of subscriptions from the manifest more explicit

:+1:

sean797 replied:
Would everyone be okay with renaming the current `state` param to `subscription_state` & adding a `manifest_state` param?

akofink replied:
I'm fine with it. Perhaps name it `pool_state` for consistency, and just `state` for the manifest state to follow what we're doing in the other ansible modules.

sean797 commented:
thanks for testing @akofink 👍  That is expected, I believe if you remove a sub from the manifest it can cause hosts to lose subs in when reimporting/refreshing it in Katello.

sean797 commented:
Okay, finally found some time :D

I've:
- Renamed the old `state` var to `pool_state`
- Added `state` var that is the manifest state.
- Renamed ` validate_certs` var to `verify_ssl` to match the other modules in this repo.

mdellweg commented on modules/redhat_manifest.py:
```diff
@@ -0,0 +1,293 @@
+            error = json.loads(info['body'])['displayMessage']
+        except:
+            error = info['msg']
+        raise Exception(error)
```
I have the expression, that this Exception might not be caught (and transformed into module.fail_json) at every invocation of fetch_portal.

sean797 commented:
Thanks @mdellweg, updated.

mdellweg commented:
I know, it might not be easy, but have you thought about automated testing for this module? There is a test_playbook...

mdellweg commented on test/test_crud.py:
```diff
@@ -16,6 +16,7 @@
     'provisioning_template',
     'ptable',
     'repository',
+    'redhat_manifest',
```
Please keep this in alphabetical order.

sean797 commented:
I've added some tests and it looks like Mr Travis is happy with the tests and hopefully your happy with the Module ordering ;) :+1:

mdellweg commented:
@sean797 Thank you for the patience.
